website/src/_posts/2022-09-3.0.md

```diff
@@ -0,0 +1,146 @@
+      // https://uppy.io/docs/transloadit/#getAssemblyOptions-file
+      //
+      // It is important to set the secret in production:
+      // https://transloadit.com/docs/topics/signature-authentication/
+      const response = await fetch('/some-endpoint')
+      return response.json()
+    },
+  })
+```
+
+## Companion 4.0
+
+### Streaming upload
+
+Streaming upload can now be enabled in Companion with Tus. This gives greatly improved upload speeds when using remote Providers and allows uploading potentially hundreds of gigabytes without needing a large server storage.
+
+With streaming upload disabled (default), the whole file will be downloaded first, and only when the download has completely finished would the upload start.
+
+```js
+// Pseudo-code:
+const providerStream = provider.getFileStream()
+const tempFilePath = 'tmpfile'
+const fileStream = fs.createWriteStream(tempFilePath)
+await pipeline(providerStream, fileStream)
+await tus.uploadFile(tempFilePath)
+```
+
+When streaming upload is enabled, Companion will start downloading the file from the provider (for example Google Drive), while at the same time starting the upload to the destination (for example Tus) at the same time, and send every chunk of data consecutively.
+
+```js
+// Pseudo-code:
+const providerStream = provider.getFileStream()
+await tus.uploadStream(providerStream)
+```
+
+For more information, [see the docs](https://uppy.io/docs/companion/).
+
+### Event emitter
+
+Companion now has a server-side event emitter that can be used to detect when uploads start, finish and fail, without having to depend on the client (so it works even if the user has closed their browser).
+
+Example code:
+
+```js
+const { app, emitter } = companion.app(options)
+
+emitter.on('upload-start', ({ token }) => {
+  console.log('Upload started', token)
+
+  function onUploadEvent ({ action, payload }) {
+    if (action === 'success') {
+      emitter.off(token, onUploadEvent) // avoid listener leak
+      console.log('Upload finished', token, payload.url)
+    } else if (action === 'error') {
+      emitter.off(token, onUploadEvent) // avoid listener leak
+      console.error('Upload failed', payload)
+    }
+  }
+  emitter.on(token, onUploadEvent)
+})
+```
+
+### `Async`/`await`-ify Companion
+
+Companion’s internal request code and Providers have been rewritten to `async`/`await` which has grossly simplified and reduced our amount of code. As can be seen [in this PR](https://github.com/transloadit/uppy/pull/3953), we lost 425 lines!
+
+For example the new facebook Provider implementation is only [100 lines of code](https://github.com/transloadit/uppy/blob/35812ca378b9b02290a2ab51261ddaeff8211fda/packages/%40uppy/companion/src/server/provider/facebook/index.js), [down from 194](https://github.com/transloadit/uppy/blob/7b0db7d3a0163374cfbde01e826c86ead4f0aacb/packages/%40uppy/companion/src/server/provider/facebook/index.js). This should make it much easier to implement new Providers.
```

Comment:
```suggestion
Companion’s internal request code and Providers have been rewritten to `async`/`await` which has grossly simplified and reduced our amount of code. This will lead to more less bugs and security issues. It will also make it easier to implement new custom Providers.

We have also replaced the deprecated [request](https://github.com/request/request) library with [got](https://github.com/sindresorhus/got). This also removed the need for [purest](https://github.com/simov/purest).
```